### PR TITLE
Terraform apply progressive logging/debug stderr capture

### PR DESCRIPTION
**Pull request by dacbd**
Closes: #938
Closes: #1039

<details>
<summary>ref terraform json logs</summary>

```
{
  '@level': 'info',
  '@message': 'Terraform 1.1.9',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:08:05.267173-07:00',
  terraform: '1.1.9',
  type: 'version',
  ui: '1.0'
}
```
```
{
  '@level': 'info',
  '@message': 'iterative_cml_runner.runner: Plan to create',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:08:05.610245-07:00',
  change: {
    resource: {
      addr: 'iterative_cml_runner.runner',
      module: '',
      resource: 'iterative_cml_runner.runner',
      implied_provider: 'iterative',
      resource_type: 'iterative_cml_runner',
      resource_name: 'runner',
      resource_key: null
    },
    action: 'create'
  },
  type: 'planned_change'
}
```

```
{
  '@level': 'info',
  '@message': 'Plan: 1 to add, 0 to change, 0 to destroy.',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:08:05.610327-07:00',
  changes: { add: 1, change: 0, remove: 0, operation: 'plan' },
  type: 'change_summary'
}
```

```
{
  '@level': 'info',
  '@message': 'iterative_cml_runner.runner: Creating...',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:08:05.662274-07:00',
  hook: {
    resource: {
      addr: 'iterative_cml_runner.runner',
      module: '',
      resource: 'iterative_cml_runner.runner',
      implied_provider: 'iterative',
      resource_type: 'iterative_cml_runner',
      resource_name: 'runner',
      resource_key: null
    },
    action: 'create'
  },
  type: 'apply_start'
}
```

```
{
  '@level': 'info',
  '@message': 'iterative_cml_runner.runner: Still creating... [1m0s elapsed]',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:09:05.669267-07:00',
  hook: {
    resource: {
      addr: 'iterative_cml_runner.runner',
      module: '',
      resource: 'iterative_cml_runner.runner',
      implied_provider: 'iterative',
      resource_type: 'iterative_cml_runner',
      resource_name: 'runner',
      resource_key: null
    },
    action: 'create',
    elapsed_seconds: 60
  },
  type: 'apply_progress'
}
```

```
{
  '@level': 'info',
  '@message': 'iterative_cml_runner.runner: Creation complete after 1m7s [id=iterative-19fbm9ljiihqh]',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:09:12.644961-07:00',
  hook: {
    resource: {
      addr: 'iterative_cml_runner.runner',
      module: '',
      resource: 'iterative_cml_runner.runner',
      implied_provider: 'iterative',
      resource_type: 'iterative_cml_runner',
      resource_name: 'runner',
      resource_key: null
    },
    action: 'create',
    id_key: 'id',
    id_value: 'iterative-19fbm9ljiihqh',
    elapsed_seconds: 67
  },
  type: 'apply_complete'
}
```

```
{
  '@level': 'info',
  '@message': 'Apply complete! Resources: 1 added, 0 changed, 0 destroyed.',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:09:12.652475-07:00',
  changes: { add: 1, change: 0, remove: 0, operation: 'apply' },
  type: 'change_summary'
}
```

```
{
  '@level': 'info',
  '@message': 'Outputs: 0',
  '@module': 'terraform.ui',
  '@timestamp': '2022-06-08T11:09:12.652506-07:00',
  outputs: {},
  type: 'outputs'
}
```

</details>

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -179,6 +181,41 @@ const gpuPresent = async () => {
...
+    tfProc.stdout.on('data', (buf) => {
+      const parse = (line) => {
+        if (line === '') return;
+        try {
+          const log = JSON.parse(line);
+          if (log['@level'] === 'error') {
+            winston.error(`terraform error: ${log['@message']}`);
+          } else {
+            winston.info(log['@message']);
+          }
```
Comment (0x2b3bfa0) [inline src/utils.js:197]: `-Wperhaps-it-is-more-clear`

```suggestion
          const { '@level': level, '@message': message } = JSON.parse(line);
          if (level === 'error') {
            winston.error(`terraform error: ${message}`);
          } else {
            winston.info(message);
          }
```

File: src/terraform.js [hunk 2]
```diff
@@ -36,8 +36,27 @@ const init = async (opts = {}) => {
...
+  if (json) {
+    const env = process.env;
+    if (env.TF_LOG_PROVIDER === undefined) env.TF_LOG_PROVIDER = 'DEBUG';
+    try {
+      await tfCapture(
+        'terraform',
+        [`-chdir='${dir}'`, 'apply', '-auto-approve', '-json'],
+        {
+          cwd: process.cwd(),
+          env: env,
```
Comment (0x2b3bfa0) [inline src/terraform.js:49]: `-Wrepetitive-although-consistent`
```suggestion
          env,
```

File: src/terraform.js [hunk 3]
```diff
@@ -36,8 +36,27 @@ const init = async (opts = {}) => {
 };
 
 const apply = async (opts = {}) => {
-  const { dir = './' } = opts;
-  return await exec(`terraform -chdir='${dir}' apply -auto-approve`);
+  const { dir = './', json = false } = opts;
+  if (json) {
+    const env = process.env;
```
Comment (0x2b3bfa0) [inline src/terraform.js:41]: `-Wynaut`

```suggestion
    const { env } = process;
```

File: src/utils.js (hunk 1 above)
Comment (dacbd) [reply src/utils.js:197]: we can get one more opinion @casperdcl / @DavidGOrtega ?

**Review by 0x2b3bfa0 — APPROVED** [review]
LMTM[^1]

[^1]: Looks mergeable to me 😄

**Comment by dacbd** [comment]
> 1. Looks mergeable to me 😄 

Savage 🤣 do you have more thoughts?

File: src/utils.js (hunk 1 above)
Comment (casperdcl) [inline src/utils.js:197]: I prefer unpacking too

File: src/utils.js [hunk 4]
```diff
@@ -179,6 +181,41 @@ const gpuPresent = async () => {
...
+      };
+      buf.toString('utf8').split('\n').forEach(parse);
+    });
+    tfProc.stderr.on('data', (buf) => {
+      stderrCollection.push(buf);
+    });
+    tfProc.on('close', (code, signal) => {
+      if (code !== 0) {
+        const stderrOutput = Buffer.concat(stderrCollection).toString('utf8');
+        reject(stderrOutput);
```
Comment (casperdcl) [inline src/utils.js:211]: `reject` vs `winston.error`?

File: src/utils.js (hunk 1 above)
Comment (DavidGOrtega) [reply src/utils.js:197]: always deestructure

File: src/terraform.js [hunk 5]
```diff
@@ -36,8 +36,25 @@ const init = async (opts = {}) => {
 };
 
 const apply = async (opts = {}) => {
-  const { dir = './' } = opts;
-  return await exec(`terraform -chdir='${dir}' apply -auto-approve`);
+  const { dir = './', json = false } = opts;
+  if (!json) return await exec(`terraform -chdir='${dir}' apply -auto-approve`);
```
Comment (DavidGOrtega) [inline src/terraform.js:40]: I do not fully understand why we keep the non json exec if we are not giving any chance to use it aside unit testing

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]

File: src/terraform.js (hunk 5 above)
Comment (DavidGOrtega) [reply src/terraform.js:40]: > we could change that one as well.

Yes, well indeed we do not have to change that... we just only need to make apply behave one way

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
what about destroy?

Im missing the destroy logs when something goes wrong and cml runner invokes destroy

**Comment by DavidGOrtega** [comment]
I do not think that this PR closes #1039 it might display the errors but it might not the error code 143 here is the key and also the lack of ANY log, we have to take into account that we wait for terraform to finish and then display but terraform never ended, indeed

**Comment by dacbd** [comment]
I missed that sorry

File: src/utils.js (hunk 4 above)
Comment (DavidGOrtega) [reply src/utils.js:211]: thats perfect @dacbd

File: src/utils.js [hunk 6]
```diff
@@ -179,6 +181,41 @@ const gpuPresent = async () => {
...
+        if (line === '') return;
+        try {
+          const { '@level': level, '@message': message } = JSON.parse(line);
+          if (level === 'error') {
+            winston.error(`terraform error: ${message}`);
+          } else {
+            winston.info(message);
+          }
+        } catch (err) {
+          // Failed to parse json from buffer
```
Comment (DavidGOrtega) [inline src/utils.js:199]: Im pretty sure that anyone can understand that JSON failed 😁

**Comment by dacbd** [comment]
> what about destroy?
> 
> Im missing the destroy logs when something goes wrong and cml runner invokes destroy

This is only called on the server?/ and we can add that as well, but perhaps just setting a tf log file for that case is better?

If apply times out (provisioning took too long) the provider calls the delete function, and apply exits with `1` and so the extra logs would be displayed.

**Comment by DavidGOrtega** [comment]
Well actually destroy does not matter because it fires and forget so its immediate. @dacbd 👍

**Review by DavidGOrtega — APPROVED** [review]
👍 lgtm! 🙏

**Comment by 0x2b3bfa0** [comment]
> _do you have more thoughts?_

None, back-in-time :shipit:

**Pull request merged**